### ISSUE TITLE
Change Request: `context.report` should accept wider type for `node`

### ISSUE THREAD

**Issue by paulius-valiunas** [eslint/eslint#19733]
### ESLint version

9.26.0

### What problem do you want to solve?

Currently, only variables of type `Node` can be used as the `node` argument for `RuleContext.report`. The `Node` type is defined like so:

```ts
export interface NodeMap {
    AssignmentProperty: AssignmentProperty;
    CatchClause: CatchClause;
    Class: Class;
    ClassBody: ClassBody;
    Expression: Expression;
    Function: Function;
    Identifier: Identifier;
    Literal: Literal;
    MethodDefinition: MethodDefinition;
    ModuleDeclaration: ModuleDeclaration;
    ModuleSpecifier: ModuleSpecifier;
    Pattern: Pattern;
    PrivateIdentifier: PrivateIdentifier;
    Program: Program;
    Property: Property;
    PropertyDefinition: PropertyDefinition;
    SpreadElement: SpreadElement;
    Statement: Statement;
    Super: Super;
    SwitchCase: SwitchCase;
    TemplateElement: TemplateElement;
    VariableDeclarator: VariableDeclarator;
}

export type Node = NodeMap[keyof NodeMap];
```

This means you cannot use it with other types of nodes, such as `Comment`.

### What do you think is the correct solution?

What `RuleContext.report` really needs is a reference to the source location of the node. These properties come from the `BaseNodeWithoutComments` interface that each of the allowed `Node` types extend. However, those are not the only types that extend it. For example:
```ts
export interface BaseNodeWithoutComments {
    // Every leaf interface that extends BaseNode must specify a type property.
    // The type property should be a string literal. For example, Identifier
    // has: `type: "Identifier"`
    type: string;
    loc?: SourceLocation | null | undefined;
    range?: [number, number] | undefined;
}
export interface Comment extends BaseNodeWithoutComments {
    type: "Line" | "Block";
    value: string;
}
```

The `Comment` type has the same location properties as any other node type on the list, but it's not accepted as an argument for `RuleContext.report`. **I think any type that extends `BaseNodeWithoutComments` should be accepted.**

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

It's worth noting that `RuleContext` is generic and the type for `Node` is `unknown` by default. However, that type is in `@eslint/core` and it's only re-exported from `eslint` with the generic parameter set to the `Node` type quoted above. As far as I know, there's no way to use it with a different node type without explicitly importing `@eslint/core`

**Comment by paulius-valiunas** [comment]
One potential solution that was proposed in the Discord server was to use `Comment.loc` as the `loc` argument instead of `node`. That didn't work because `BaseNodeWithoutComments.loc` is actually nullable. I don't know what eslint does in case it's undefined, but I'd guess there's already some handling for that, which could apply to other node types as well.

So, an alternative solution would be to make the `loc` argument for `RuleContext.report` nullable too. But that might cause people to simply forget to add a location at all. I think it's better to be explicit about it, even if the node being reported doesn't have a location.

**Comment by fasttime** [comment]
Thanks for the issue @paulius-valiunas. Could you please show an example of TypeScript code that fails to compile when `context.report()` is passed valid arguments, so that we can be sure how the fix should have to look like?

**Comment by paulius-valiunas** [comment]
@fasttime sure, here's a minimal example:
```ts
  create(context) {
    return {
      Program(node) {
        for (const comment of node.comments ?? []) {
          let found = 0;
          context.report({
            // @ts-expect-error -- comments are not considered nodes but this still works at runtime
            node: comment,
            messageId: "foo",
          });
        }
      },
    };
  }
```

**Comment by fasttime** [comment]
> ```ts
>   create(context) {
>     return {
>       Program(node) {
>         for (const comment of node.comments ?? []) {
>           context.report({
>             // @ts-expect-error -- comments are not considered nodes but this still works at runtime
>             node: comment,
>             messageId: "foo",
>           });
>         }
>       },
>     };
>   }
> ```
> 
> It produces the following error:
> 
> ```
> Type 'Comment' is not assignable to type 'Node'.
>   Property 'tail' is missing in type 'Comment' but required in type 'TemplateElement'.ts(2322)
> index.d.ts(545, 5): 'tail' is declared here.
> ```

There are no explicit types in your example. Are you typing your rule object with `Rule.RuleModule` like in the snippet below, or are you using a helper function to generate the rule object?

```ts
const rule: Rule.RuleModule = {
  create(context) {
    ...
  }
};
```

**Comment by paulius-valiunas** [comment]
Sorry I skipped that part. I wasn't using a helper, my rule looks exactly like in your example.

**Comment by fasttime** [comment]
Okay, thanks. Using `Rule.RuleModule` is not recommended for new rules but I'm getting the same error with `JSRuleDefinition`.

The documentation for `context.report` indicates that the `node` option, if present, should be an AST node. The option name "node" only reinforces this assumption. Despite that, ESLint itself doesn't follow this specification, and many built-in rules do fill the `node` property with a comment or token in order to set the violation range. An example is the `no-fallthrough` rule:

https://github.com/eslint/eslint/blob/46eea6d1cbed41d020cb76841ebba30710b0afd0/lib/rules/no-fallthrough.js#L225-L228

or `no-async-promise-executor`:

https://github.com/eslint/eslint/blob/46eea6d1cbed41d020cb76841ebba30710b0afd0/lib/rules/no-async-promise-executor.js#L35-L41

Changing this behavior would be a breaking change, so I think we should rather treat this as a bug in the documentation and types, and update both.

**Comment by paulius-valiunas** [comment]
Well, comments are part of the AST and `Comment` does extend `BaseNodeWithoutComments`, so I'd argue that it _is_ a node, it's just not on the short list that the `Node` type alias includes.

**Comment by fasttime** [comment]
While comments could be considered an extension of AST nodes, tokes are not. A possible fix could look like this:

* Update `JSRuleDefinition` with a more generic type for [`Node`](https://github.com/eslint/eslint/blob/20a9e59438fde3642ab058cc55ee1b9fa02b6391/lib/types/index.d.ts#L1278) that accepts `ESTree.Comment` and `AST.Token` along with `ESTree.Node`.
* Apply the same fix to the [`Node`](https://github.com/eslint/eslint/blob/20a9e59438fde3642ab058cc55ee1b9fa02b6391/lib/types/index.d.ts#L1180) option of `Rule.RuleModule` for consistency. This should not break compatibility, but I haven't test that yet.
* Clarify the documentation of `context.report` to state that AST nodes are not the only valid syntax elements for the [`node` option](https://github.com/eslint/eslint/blob/20a9e59438fde3642ab058cc55ee1b9fa02b6391/docs/src/extend/custom-rules.md?plain=1#L166).
* The docs for the deprecated property [`LintReport#nodeType`](https://github.com/eslint/eslint/blob/20a9e59438fde3642ab058cc55ee1b9fa02b6391/docs/src/integrate/nodejs-api.md?plain=1#L675) already indicate that this is the type of a node or token. We could extend it to specify that it could be also the type of a comment.

Curious to know what the rest of the team thinks. @eslint/eslint-team

**Comment by paulius-valiunas** [comment]
To be honest, I find the distinction between "token" and "node" a bit ambiguous. In my understanding, source code is initially parsed into a flat list of tokens, but once everything is put in an AST, only nodes are left. For example, a for loop would first be parsed into tokens "for", "(", etc., but you won't find them in the AST, you'll only find `init`, `test`, `update` and `body`, which are all nodes. But this is just my limited understanding and I might be wrong, or you might be using different nomenclature at ESLint.

Either way, instead of explicitly adding `Comment` to the list of accepted types, wouldn't it be more universal and inclusive to accept `BaseNodeWithoutComments` instead of `Node`? That would also include nodes from `typescript-estree` and other parsers (at least the ones that are often used with ESLint) that usually try to extend estree as much as possible, as well as any other nodes/tokens that are not included in the `Node` alias.

**Comment by mdjermanovic** [comment]
> So maybe we should just make a new type with `type` and `loc` and set that as the expected type anywhere this value is used?

I think this makes sense, but just to note that the `loc` property is needed for JS language. For other languages, this object can be anything that can be passed to that language's `SourceCode#getLoc()`, i.e, it doesn't necessarily need to have a property named `loc`.

https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/linter/report-translator.js#L404

**Comment by nzakas** [comment]
Yes, thanks for clarifying. I meant to say for `JSRuleDefinition`, specifically. We pass in the type that `report()` accepts here:
https://github.com/eslint/eslint/blob/main/lib/types/index.d.ts#L1180

Each language is responsible for ensuring that type is correct.

**Comment by fasttime** [comment]
Sounds good to create a new type with just `type` and `loc` for the `Node` option of `JSRuleDefinition`, and if possible for `Rule.RuleModule` (including `Rule.RuleContext`). Marking as accepted.

Noting that we will have a discrepancy with how `eslint/css` and `eslint/json` define the `Node` option in their rule definitions. In these plugins, `Node` is defined as a discrete union of known node types (similarly to how `ESTree.Node` is defined) rather than as a general base type. But since other language will not be affected by this change, we can always consider aligning the types later.

**Comment by nzakas** [comment]
I think it's okay for their to be discrepancies across language plugins as this behavior is language-specific. If there are problems with the other plugins we can address them as they're reported.

**Comment by paulius-valiunas** [comment]
> (As a note: This is part of why I wanted to deprecate the node argument a while back and encourage people to use the loc argument instead.)

That would make sense, but `loc` is nullable on most nodes so that parameter would have to be optional. I think there's a risk of people simply forgetting to add it, since Typescript wouldn't complain.

**Comment by paulius-valiunas** [comment]
Interesting... So how do you guarantee that custom parsers will always add `loc` to their AST? What if they don't?

It's kind of ugly, but you could try using declaration merging to make it mandatory in ESTree.
